### PR TITLE
test: document testlib functions (part one of X)

### PR DESCRIPTION
**Pull request by jelly**
@allisonkarlitskaya  @martinpitt @marusak what do you think of using sphinx for generating documentation of our testlib combined with some documentation on how to set up the tests. The documentation should be aimed for us but mostly for people new to writing tests.

Secondly which docstring format do we go for, the Rst style looks the best in sphinx.

https://realpython.com/documenting-python-code/#docstring-formats

Oh and we might want to make a cockpit theme for sphinx?

### CODE REVIEW

**Comment by allisonkarlitskaya** [comment]
My first take is that I'm not a huge fan.  I'm all for writing better docstrings, but I'm not sure we should be aiming to generate HTML for this stuff...

**Comment by marusak** [comment]
Can you generate the html and show how it would look like?

I am very much in favor of documenting our testlib - we expect everyone to use this to write tests yet we are fully aware it is very difficult for anyone new to understand how it works and what function they can use (I was asked so many times questions like "How do I check if this element is on the page", "How do I enter some text into input"...) so having some page (ideally in https://cockpit-project.org/documentation.html) would be awesome.

**Comment by jelly** [comment]
> Can you generate the html and show how it would look like?

image


> I am very much in favor of documenting our testlib - we expect everyone to use this to write tests yet we are fully aware it is very difficult for anyone new to understand how it works and what function they can use (I was asked so many times questions like "How do I check if this element is on the page", "How do I enter some text into input"...) so having some page (ideally in https://cockpit-project.org/documentation.html) would be awesome.

Agree!

**Comment by jelly** [comment]
> My first take is that I'm not a huge fan. I'm all for writing better docstrings, but I'm not sure we should be aiming to generate HTML for this stuff...

My plan is to first focus on adding the docstrings to everywhere it's missing, for that we need to agree on the docstrings format.

I still think having generated HTML is an added benefit, we document the [javascript API](https://cockpit-project.org/guide/latest/api-base1.html) so why not our browser API as well. The only downside is that it changes more and it's not a really strict API.

**Comment by martinpitt** [comment]
I fully agree to adding docstrings to all of testlib's public API. That can then easily be viewed with `pydoc` or the usual `help()`, and it's even fine IMHO to just simply look at the file.

I'm less excited about doing the sphinx build -- we shouldn't publish that on our website. Every project uses its own tag/version of testlib, so this is prone to version confusion. It also adds complexity to our build system, and a new build dependency. I.e. this PR doesn't test the doc build *at all*.

Also, HTML is a bit "meh", Markdown would be nicer -- but that's just a little whim, as long as it's automated I don't mind much.

Thank you!

**Comment by jelly** [comment]
- [ ] figure out an alternative easier html generation tool (with less makefile's n stuff) https://pypi.org/project/pdoc3/ maybe
- [ ] document
- [ ] is there a linter for pydoc strings? https://pypi.org/project/flake8-rst-docstrings/ ?

**Comment by jelly** [comment]
> * [ ]  figure out an alternative easier html generation tool (with less makefile's n stuff) https://pypi.org/project/pdoc3/ maybe
> * [ ]  document
> * [ ]  is there a linter for pydoc strings? https://pypi.org/project/flake8-rst-docstrings/ ?

Good candidate seems to be https://github.com/PyCQA/pydocstyle

File: test/common/testlib.py [hunk 1]
```diff
@@ -215,19 +216,39 @@ def open(self, href, cookie=None, tls=False):
...
         self.wait_js_cond("ph_select('iframe.container-frame').every(function (e) { return e.getAttribute('data-loaded'); })")
         self.cdp.invoke("reloadPageAndWait", ignoreCache=ignore_cache)
 
         self.machine.allow_restart_journal_messages()
 
     def switch_to_frame(self, name):
+        """Switch to frame in browser tab
+
+        Each page has a main frame and can have multiple subframes for example
+        iframes.
```
Comment (martinpitt) [inline test/common/testlib.py:241]: "subframes, usually iframes.

File: test/common/testlib.py [hunk 2]
```diff
@@ -271,9 +299,18 @@ def eval_js(self, code, no_trace=False):
         return None
 
     def call_js_func(self, func, *args):
+        """Call a JavaScript function
+
+        :param func: JavaScript function to call
+        :param args: arguments for the JavaScript function
+        """
```
Comment (martinpitt) [inline test/common/testlib.py:306]: This is rather stub-y -- this does not really explain much beyond the signature. But we can build on this :-)

File: test/common/testlib.py [hunk 3]
```diff
@@ -271,9 +299,18 @@ def eval_js(self, code, no_trace=False):
...
         return self.eval_js("%s(%s)" % (func, ','.join(map(jsquote, args))))
 
     def cookie(self, name):
+        """Retrieve the browsers cookie by name
```
Comment (martinpitt) [inline test/common/testlib.py:310]: "a browser cookie

File: test/common/testlib.py [hunk 4]
```diff
@@ -143,6 +143,9 @@ def attach(filename, move=False):
 
     By default the file gets copied. You can set move=True for dynamically generated
     files which are not touched by parallel tests.
+
+    :param filename: file to put in attachments directory
+    :param move: move file to attachments directory (default False)
```
[the remark this answers is not in the document]
Comment (jelly) [reply test/common/testlib.py]: Okido. done

File: test/common/testlib.py [hunk 5]
```diff
@@ -141,11 +139,12 @@
 
 
 def attach(filename, move=False):
-    '''Put a file into the attachments directory
+    """Put a file into the attachments directory.
```
Comment (martinpitt) [inline test/common/testlib.py:142]: For next time: No periods in short description.

**Review by martinpitt — APPROVED** [review]
Thanks!

**Pull request merged**